### ISSUE TITLE
UTF-8 characters in trusted/allowed file names

### ISSUE THREAD

**Issue by vahldiek** [oscarlab/graphene#1337]
Hi,

Just ran into an issue with UTF-8 encoded files names (i.e., file names with ü) when used in the manifest to create trusted/allowed files. To recreate you'd have to create a file with such a character, include it as a trusted file in your favorite manifest, then run the pal-sgx-signer (which will break), token generation and then the loader (which will break again).

I've attached an updated version of the [pal_sgx_sign.py.txt](https://github.com/oscarlab/graphene/files/4222673/pal_sgx_sign.py.txt) that includes the required *.encode calls and encoding="utf-8" additions within python. I've not dug further into the manifest initialization within the libOS when loading an application, but I expect the code to be equally unprepared.

This issue is mainly to document my findings.

**Comment by mkow** [comment]
Thanks! Could you create a PR with the mentioned fixes?

**Comment by woju** [comment]
I posted this comment in #1355 (here with minor fixes):

Filenames are not required to be valid UTF-8 either, so hardcoding `.encode('utf-8')` etc is wrong. In principle, filenames are sequences of bytes (octets) and the only requirement is that they do not contain U+0 (`\0`) and U+2F (`/`) characters, because those are separators for C string and path, respectively. (Filesystems may impose limits on total path length, which they do at about 256 octets or slightly less, but that's implementation, not API).

Python has two mechanisms to cope with that:
- https://docs.python.org/library/os#os.PathLike and https://docs.python.org/library/os#os.fspath
- `errors='surrogateescape'` on paths (https://docs.python.org/library/sys#sys.getfilesystemencodeerrors).

Both of those mechanisms are since Python 3.6, but we support xenial (16.04 LTS), which has Python 3.5, which is a problem. On Python<=3.5 argument to `open()` has to be `str`, which implies that it is correctly encoded.

If we'd like to fix this correctly, we should allow for arbitrary byte sequences in paths. We could:
- specify that manifest must be encoded in UTF-8 (and open it with `encoding='utf-8');
- allow `\`-escapes or some other way to specify arbitrary bytes;
- tell people that incorrectly encoded file paths (wrt to current locale) are unsupported on python3.5/xenial.

To fix the path encoding problems on <=3.5, I usually do something like this:

```python
try:
    from os import fspath
except ImportError:
    # python<3.6
    fspath = str
```

Then one can use also `bytes` and `pathlib.Path` as file paths:
```python
with open(fspath(filepath)) as file:
    ...
```

Which fixes at least non-ASCII paths in system locale. Other locales are not fixable.

**Comment by dimakuv** [comment]
I don't think this is a problem anymore, since we use TOML for our manifest files, and TOML requires UTF-8 encoding. @mkow You changed our Python scripts to TOML, do you think this issue still relevant? I think not, so closing for now. But feel free to re-open.

**Comment by mkow** [comment]
I think it's still relevant, because as @woju mentioned, file paths doesn't need to be in UTF-8 on some platforms. We should at least add a test to verify that this works (I think it doesn't).
